### PR TITLE
MGMT-13442 add generic issue filtering for triage

### PR DESCRIPTION
**Pull request by paul-maidment**
The aim of this PR is twofold:
* reorganize code so it's much easier to
  * run dryrun mode
  * filter issues by custom criteria
* add filtering by labels
* tests added

### CODE REVIEW

File: triage_resolving_filters.json
Comment (rccrdpccl) [inline triage_resolving_filters.json:4]: AFAIU this message will be looked for in the comments for closing. Is this sitll necessary? Should we get instead the message from the signature class, as [we already have that](https://github.com/openshift-assisted/assisted-installer-deployment/blob/master/tools/triage/add_triage_signature.py#L330)?

File: Makefile
Comment (rccrdpccl) [inline Makefile:44]: There a few warnings of deprecation of some libraries, we can get it resolved after closing this

File: dev-requirements.txt
Comment (rccrdpccl) [inline dev-requirements.txt:15]: Needed for `match` statement

File: tools/jira_client/jira_dryrun_client.py
Comment (osherdp) [inline tools/jira_client/jira_dryrun_client.py:23]: per above comment, it should be possible to just use status names instead of IDs, so this can be simplified

File: tools/jira_client/jira_api.py
Comment (osherdp) [inline tools/jira_client/jira_api.py:13]:
```suggestion
        self._logger.info("Linking %s as related to %s", issue.key, root_issue.key)
```

**Comment by rccrdpccl** [comment]
@osherdp can we test this in Jenkins before merging?

File: tools/jira_client/jira_api.py
Comment (osherdp) [inline tools/jira_client/jira_api.py:17]: sentence a bit weird, maybe just
```suggestion
        self._logger.info("Closing issue %s", issue)
```

File: tools/jira_client/jira_api.py
Comment (osherdp) [inline tools/jira_client/jira_api.py:27]: why should it be none?

File: tools/jira_client/jira_api.py
Comment (osherdp) [inline tools/jira_client/jira_api.py:34]: I think that ultimately after 3 retries we should raise the error if we're consistently getting one
Otherwise who knows what potential issues we might miss

File: tools/triage/filters.py
Comment (osherdp) [inline tools/triage/filters.py:178]: I prefer the top level function to create the relevant filters (either from args or from json or from any other source) and provide them to a function of ``Filters.close_issues_by_filters`` to actually do the closing

File: tools/triage/filters.py
Comment (osherdp) [inline tools/triage/filters.py:28]:
```suggestion
        return issue.fields.status.name not in ("Closed", "Done", "Obsolete")
```

**Review by osherdp — CHANGES_REQUESTED** [review]
partial review

**Comment by osherdp** [comment]
> @osherdp can we test this in Jenkins before merging?

yes, it's possible 
I'll send you the details in PM

File: tools/jira_client/jira_client_factory.py
Comment (paul-maidment) [reply tools/jira_client/jira_client_factory.py:12]: We need to pass in a client into the dryrun client, this is why we don't just return immediately.

File: tools/jira_client/jira_api.py
Comment (paul-maidment) [reply tools/jira_client/jira_api.py:27]: We kept the existing logic exactly as it was to avoid issues, we can change it if desired?

File: tools/jira_client/jira_api.py
Comment (paul-maidment) [reply tools/jira_client/jira_api.py:34]: We kept the existing logic exactly as it was to avoid issues, we can change it if desired?

Comment (osherdp) [reply tools/jira_client/jira_api.py:34]: now that it's ``jira_api`` it becomes more desirable to have the basics right
it makes this code more reusable than it was previously, so we wouldn't want people to rely on something that hides defects in it

File: tools/jira_client/jira_dryrun_client.py
Comment (osherdp) [inline tools/jira_client/jira_dryrun_client.py:31]: why masking the status? I think it can just be:
```suggestion
        print(
            f"Transitioned issue key={issue.issue.key} to {target_transition_id},
            file=self._dry_run_stream,
            flush=True,
        )
```

File: tools/triage/add_triage_signature.py
Comment (osherdp) [inline tools/triage/add_triage_signature.py:21]: the convention we (mostly) follow for imports is splitting to:
1. built-ins
2. third-party libraries
3. internal code

So this should be right beside
```
from tools.triage.common import get_cluster_logs_base_url
```

(this applies here and elsewhere)

File: tools/triage/filters.py
Comment (osherdp) [inline tools/triage/filters.py:69]: pretty much does the same
```suggestion
        return SIGNATURE_BY_TYPE[signature_class_alias_name]
```

File: tools/triage/filters.py
Comment (osherdp) [inline tools/triage/filters.py:103]:
```suggestion
                raise ValueError(f"unknown filter type {other}")
```

File: tools/triage/filters.py
Comment (osherdp) [inline tools/triage/filters.py:111]:
```suggestion
```
delete. this should be dead-code

File: tests/tools/jira/test_jira_api.py
Comment (osherdp) [inline tests/tools/jira/test_jira_api.py:10]:
```suggestion
    return Mock(key=key)
```
maybe you don't even need this function as it's much shorter with its body

File: tests/tools/jira/test_jira_api.py
Comment (osherdp) [inline tests/tools/jira/test_jira_api.py:16]:
```suggestion
```
all methods of a mock should automatically generate a mock object

File: tests/tools/jira/test_jira_api.py
Comment (osherdp) [inline tests/tools/jira/test_jira_api.py:49]: I see it's reoccurring in the same way, so you can do:
```python
from pytest import fixture

@pytest.fixture()
def jira_api():
    return JiraAPI(jira_client=Mock(), logger=Mock())

def test_should_call_underlying_jira_client_to_link_issue_to_root_issue(jira_api):
    issue = get_mock_issue("mock_issue")
    root_issue = get_mock_issue("root_issue")
    jira_api.link_issue_to_root_issue(issue=issue, root_issue=root_issue)
    jira_api._jira_client.create_issue_link.assert_called_once_with("relates to", "mock_issue", "root_issue")
```

File: tests/tools/jira/test_jira_client_factory.py
Comment (osherdp) [inline tests/tools/jira/test_jira_client_factory.py:12]:
```suggestion
    assert isinstance(jira_client, jira.client.JIRA)
```

File: tests/tools/jira/test_jira_dryrun_client.py
Comment (osherdp) [inline tests/tools/jira/test_jira_dryrun_client.py:9]:
```suggestion
@pytest.fixture()
def devnull_stream():
    with open(os.devnull, "w") as devnull:
        yield devnull
```

then tests can just have it as an argument. e.g.:
```python
def test_something(devnull):
...
```

File: tests/tools/jira/test_jira_dryrun_client.py
Comment (osherdp) [inline tests/tools/jira/test_jira_dryrun_client.py:18]:
```suggestion
class MockAttributeErrorJiraClient:
```

File: tests/tools/jira/test_jira_dryrun_client.py
Comment (osherdp) [inline tests/tools/jira/test_jira_dryrun_client.py:19]: it's rather unused, no? so why do we need it for the tests?

File: tests/tools/jira/test_jira_dryrun_client.py
Comment (osherdp) [inline tests/tools/jira/test_jira_dryrun_client.py:33]: just a second, the dry-run client is able to proxy any call to jira client? how can we make sure it doesn't do write actions through those calls?

File: tools/triage/filters.py
Comment (osherdp) [inline tools/triage/filters.py:48]:
```suggestion
            if self.identifier in issue.fields.labels:
                return True
```

File: tests/tools/triage/test_filters.py
Comment (osherdp) [inline tests/tools/triage/test_filters.py:38]: use ``isinstance`` here and elsewhere

File: tests/tools/triage/test_filters.py
Comment (osherdp) [inline tests/tools/triage/test_filters.py:58]:
```suggestion
        Filters.create_filter(jira_client, "invalid-type", "whatever", root_issue_key, "my message")
```

File: tests/tools/triage/test_filters.py
Comment (osherdp) [inline tests/tools/triage/test_filters.py:57]:
```suggestion
    with pytest.raises(Exception, match="unknown filter type invalid-type"):
```

**Review by osherdp — CHANGES_REQUESTED** [review]

File: tools/jira_client/jira_dryrun_client.py
Comment (rccrdpccl) [reply tools/jira_client/jira_dryrun_client.py:31]: Yeah, you are right, leftover from status_id and blindly replaced :sweat:

File: tests/tools/jira/test_jira_dryrun_client.py
Comment (rccrdpccl) [reply tests/tools/jira/test_jira_dryrun_client.py:33]: It's the humans responsibility to add all the write methods to the JiraDryRun (just as we hide the write call behind an `if not dry run`, but automatically all read methods will work without the need to add them as boilerplate.

The alternative I see is to add _all_ the needed methods (read and write), and proxy the read ones manually.

The risk is that users will think this is a magic class and just using without adding write methods that might be used.

WDYT?

File: tools/triage/add_triage_signature.py
Comment (rccrdpccl) [reply tools/triage/add_triage_signature.py:21]: One day we'll have this file to pass pylint :sweat_smile:

**Comment by rccrdpccl** [comment]
/hold
holding it because we need to revert a small change that I am using for testing in Jenkins

File: tests/tools/jira/test_jira_dryrun_client.py
Comment (osherdp) [reply tests/tools/jira/test_jira_dryrun_client.py]: Yeah, I guess it's not that easy to change
Ultimately we can define an interface with ``abc`` to make sure all required methods are implemented. but I get how it's too much code
I also thought about a read-only jira tokens, but it requires us to create a new user because there's no read-only scope we can switch to

All in all, I think it's fine as-is

Comment (rccrdpccl) [reply tests/tools/jira/test_jira_dryrun_client.py]: Got it, good point about the read token, we can keep it in mind for future iterations as I am aware this can potentially be dangerous.

**Comment by rccrdpccl** [comment]
@osherdp I believe all comments are addressed now, we are ready when you are :)

File: tests/tools/jira/test_jira_dryrun_client.py
Comment (osherdp) [inline tests/tools/jira/test_jira_dryrun_client.py:14]: apply same suggestion as in a different case (``Mock(key=key)``)
and let's just get rid of those ``get_mock_issue`` functions. they are redundant and can be replaced with the body of the function

Comment (osherdp) [reply tests/tools/jira/test_jira_dryrun_client.py:14]: this applies here and elsewhere, unless the function is more complex like in ``get_mock_issue``

File: tests/tools/triage/test_filters.py
Comment (osherdp) [inline tests/tools/triage/test_filters.py:44]:
```suggestion
    with pytest.raises(ValueError, match="unknown filter type invalid-type"):
```

File: tests/tools/triage/test_filters.py
Comment (osherdp) [inline tests/tools/triage/test_filters.py:119]:
```suggestion
    filters = [
        LabelFilter(
            identifier="example_label",
            root_issue=None,
            message="this is another example message",
        )
    ]
```

Comment (osherdp) [reply tests/tools/triage/test_filters.py:119]: same goes for ``issues``. less mutations is better

File: tools/triage/add_triage_signature.py
Comment (osherdp) [reply tools/triage/add_triage_signature.py]: lot's of places still have this. can you have another look?

**Review by osherdp — CHANGES_REQUESTED** [review]

File: tools/triage/add_triage_signature.py
Comment (rccrdpccl) [reply tools/triage/add_triage_signature.py]: :thinking:  it might be me but I don't see any out of order? Are you referring to this, right?

File: tests/tools/jira/test_jira_api.py
Comment (osherdp) [inline tests/tools/jira/test_jira_api.py:3]: imports out of order

File: tests/tools/jira/test_jira_client_factory.py
Comment (osherdp) [inline tests/tools/jira/test_jira_client_factory.py:6]: should be:
```suggestion
import sys

import jira

from tools.jira_client import (
   JiraClientFactory,
   JiraDryRunClient
)
```

File: tests/tools/jira/test_jira_dryrun_client.py
Comment (osherdp) [inline tests/tools/jira/test_jira_dryrun_client.py:3]:
```suggestion
import os
from unittest.mock import Mock

import pytest
```

File: tests/tools/triage/test_filters.py
Comment (osherdp) [inline tests/tools/triage/test_filters.py:4]: here as well, should be:
```suggestion
import os
from unittest.mock import Mock, call

import pytest
```

File: tools/jira_client/jira_client_factory.py
Comment (osherdp) [inline tools/jira_client/jira_client_factory.py:4]:
```suggestion
import jira

from tools import consts
from .jira_dryrun_client import JiraDryRunClient
```

File: tools/triage_status_report.py
Comment (osherdp) [inline tools/triage_status_report.py:10]:
```suggestion
import json
import os
import sys
from typing import List
from urllib import parse

import requests

from tools.jira_client import JiraClientFactory
```

File: tools/triage/add_triage_signature.py
Comment (osherdp) [reply tools/triage/add_triage_signature.py]: adding some comments for those places (not for all though)

**Review by osherdp — CHANGES_REQUESTED** [review]

File: tools/triage/add_triage_signature.py
Comment (rccrdpccl) [reply tools/triage/add_triage_signature.py]: Oh, thanks for pointing out, I was focusing only on the order of our packages :facepalm: 

Added `isort` so this can be done by a machine - which is much better than me at this job :grin:

**Review by osherdp — APPROVED** [review]
Very nice. Thank you!

File: tools/triage/add_triage_signature.py
Comment (osherdp) [reply tools/triage/add_triage_signature.py]: Yes, that's a great idea
Thank you!

**Review by osherdp — APPROVED** [review]

**Pull request merged**
